contracts/oracle/DynamicVoltSystemOracle.sol

```diff
@@ -0,0 +1,205 @@
+        );
+        oraclePrice = currentOraclePrice;
+        periodStartTime = _periodStartTime;
+    }
+
+    // ----------- Getters -----------
+
+    /// @notice Get the current scaled oracle price.
+    /// Applies the change rate linearly over {TIMEFRAME}.
+    /// The value is expressed with 18 decimals.
+    // prettier-ignore
+    function getCurrentOraclePrice() public view returns (uint256) {
+        uint256 cachedStartTime = periodStartTime; /// save a single warm SLOAD if condition is false
+        if (cachedStartTime >= block.timestamp) { /// only accrue interest after start time
+            return oraclePrice;
+        }
+
+        uint256 cachedOraclePrice = oraclePrice; /// save a single warm SLOAD by using the stack
+        uint256 timeDelta = Math.min(block.timestamp - cachedStartTime, TIMEFRAME);
+        uint256 periodPriceChange = cachedOraclePrice * actualChangeRate;
+        uint256 priceDelta = periodPriceChange * timeDelta / TIMEFRAME / 1e18;
+
+        return cachedOraclePrice + priceDelta;
+    }
+
+    /// ------------- Governor Only API -------------
+
+    /// @notice Set the base VOLT rate.
+    /// This will also refresh the actual VOLT rate.
+    /// Callable only by governor.
+    /// @param newBaseChangeRate new value for the baseChangeRate.
+    function updateBaseRate(uint256 newBaseChangeRate) external onlyGovernor {
+        /// first, compound interest
+        _compoundInterest();
+
+        uint256 oldBaseChangeRate = baseChangeRate; // SLOAD
+        baseChangeRate = newBaseChangeRate; // SSTORE
+        emit BaseRateUpdated(
+            block.timestamp,
+            oldBaseChangeRate,
+            newBaseChangeRate
+        );
+
+        /// if too few liquid reserves, adjust rate up
+        uint256 newActualChangeRate = DynamicVoltRateModel(rateModel).getRate(
+            newBaseChangeRate,
+            PCVOracle(pcvOracle).getLiquidVenuePercentage()
+        );
+        uint256 oldActualChangeRate = actualChangeRate; // SLOAD
+        actualChangeRate = newActualChangeRate; // SSTORE
+        emit ActualRateUpdated(
+            block.timestamp,
+            oldActualChangeRate,
+            newActualChangeRate
+        );
+    }
+
+    /// @notice Set the reference to the PCV oracle.
+    /// Callable only by governor.
+    /// @param newPcvOracle address of the new pcv oracle.
+    function setPcvOracle(address newPcvOracle) external onlyGovernor {
+        address oldPcvOracle = pcvOracle; // SLOAD
+        pcvOracle = newPcvOracle; // SSTORE
+        emit PCVOracleUpdated(block.timestamp, oldPcvOracle, newPcvOracle);
+    }
+
+    /// ------------- PCV Oracle Only API -------------
+
+    /// @notice Only callable by the PCV Oracle, updates the actual rate.
+    /// @param liquidPercentage the percentage of PCV that is liquid,
+    /// expressed with 18 decimals.
+    function updateActualRate(uint256 liquidPercentage) external {
```

Comment:
add a modifier here that asserts the global reentrancy lock is true, because that's the only time this function should be able to be called.